src/Disks/ObjectStorages/DiskObjectStorage.cpp

```diff
@@ -373,9 +376,14 @@ void DiskObjectStorage::removeSharedFileIfExists(const String & path, bool delet
 void DiskObjectStorage::removeSharedRecursive(
     const String & path, bool keep_all_batch_data, const NameSet & file_names_remove_metadata_only)
 {
-    auto transaction = createObjectStorageTransaction();
-    transaction->removeSharedRecursive(path, keep_all_batch_data, file_names_remove_metadata_only);
-    transaction->commit();
+    /// At most remove_shared_recursive_file_limit files are removed in one transaction
+    /// Retry until all of them are removed
+    while (exists(path))
```

Comment:
Actually this loop looks suspicious to me. It is less safe than the old version.
For example, we have a concurrent action that re-creates a dir/file constantly, in this case we can get an infinite loop. Or for some hypothetical reason dir/file is not deleted (without raising exception) - an infinite loop.

It makes sense to preserve semantic and place traversing loop here, accumulating batches and executing remove transactions during it. 

Maybe it's just a quibble).